File: math/perfect_number.rb
# @param {Integer} num
# @return {Boolean}
def check_perfect_number(num)
    return false if num <= 1
  
    divisors = [1]
    (2..Math.sqrt(num).to_i).each do |n|
        if num % n == 0
            divisors << n
            divisors << num / n
        end
    end

    divisors.reduce(:+) == num
end

File: math/perfect_number.py
class Solution:
    def checkPerfectNumber(self, num: int) -> bool:
        if num <= 1:
            return False
        
        divisors = [1]
        
        # Only check up to sqrt(num)
        for n in range(2, int(num**0.5) + 1):
            if num % n == 0:
                divisors.append(n)
                divisors.append(num // n)
        
        return sum(divisors) == num      